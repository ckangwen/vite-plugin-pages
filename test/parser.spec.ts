import { resolve } from 'path'
import { getRouteBlock } from '../src/customBlock'
import { resolveOptions } from '../src/options'

const options = resolveOptions({})

describe('Parser', () => {
  test('custom block', async() => {
    const path = resolve('./examples/vue/src/pages/blog/[id].vue')
    const routeBlock = await getRouteBlock(path, options)
    expect(routeBlock).toMatchSnapshot()
  })

  test('jsx yaml comment', async() => {
    const path = resolve('./examples/vue/src/pages/jsx.jsx')
    const routeBlock = await getRouteBlock(path, options)
    expect(routeBlock).toMatchSnapshot()
  })

  describe('option:resolveRouteBlock', () => {
    const userOptions = resolveOptions({
      resolveRouteBlock(content, filePath) {
        const isJsx = filePath.endsWith('.jsx') || filePath.endsWith('.tsx')
        if (isJsx && content.includes('definePageRoute')) {
          const RE = /definePageRoute\(([\s\S]*?)\);?/g
          const body = content.match(RE)?.[0]
          if (body) {
            // eslint-disable-next-line no-new-func
            const routeConfig = new Function(`function definePageRoute(config) { return config }; return ${body}`)()
            return routeConfig
          }
        }
      },
    })

    test('meet the condition of resolveRouteBlock', async() => {
      const path = resolve(__dirname, './data/CustomRouteBlock.jsx')
      const routeBlock = await getRouteBlock(path, userOptions)
      expect(routeBlock).toMatchSnapshot()
    })

    test('does not meet the condition of resolveRouteBlock', async() => {
      const vueFilePath = resolve('./examples/vue/src/pages/blog/[id].vue')
      const vueRouteBlock = await getRouteBlock(vueFilePath, userOptions)
      expect(vueRouteBlock).toMatchSnapshot()

      const jsxFilePath = resolve('./examples/vue/src/pages/jsx.jsx')
      const jsxRouteBlock = await getRouteBlock(jsxFilePath, userOptions)
      expect(jsxRouteBlock).toMatchSnapshot()
    })
  })
})
