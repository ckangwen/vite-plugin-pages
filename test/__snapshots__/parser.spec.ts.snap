// Vitest Snapshot v1

exports[`Parser > custom block 1`] = `
{
  "meta": {
    "requiresAuth": false,
  },
  "name": "blog-id",
}
`;

exports[`Parser > jsx yaml comment 1`] = `
{
  "meta": {
    "id": 1234,
    "requiresAuth": false,
  },
  "name": "blog-id",
}
`;

exports[`Parser > option:resolveRouteBlock > does not meet the condition of resolveRouteBlock 1`] = `
{
  "meta": {
    "requiresAuth": false,
  },
  "name": "blog-id",
}
`;

exports[`Parser > option:resolveRouteBlock > does not meet the condition of resolveRouteBlock 2`] = `
{
  "meta": {
    "id": 1234,
    "requiresAuth": false,
  },
  "name": "blog-id",
}
`;

exports[`Parser > option:resolveRouteBlock > meet the condition of resolveRouteBlock 1`] = `
{
  "meta": {
    "title": "CustomRouteBlock",
  },
}
`;
